sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -497,6 +498,11 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       StructType(schema.filterNot(_.name == parsedOptions.columnNameOfCorruptRecord))
 
     val linesWithoutHeader: RDD[String] = maybeFirstLine.map { firstLine =>
+      if (!parsedOptions.enforceSchema) {
+        CSVDataSource.checkHeader(firstLine, new CsvParser(parsedOptions.asParserSettings),
```

Comment:
I mean we could, for example, make a dataset from spark.read.text("tmp/*.csv"), preprocess it and then convert it via spark.read.csv(dataset). In this case, every file would have the header. This doesn't validate each file's header.

Shall we document this if it's hard to fix?

Reply:
Thank you for explaining of the use case. I didn't realize that the input dataset can contain multiple headers. So, potentially the headers can contain column names in different order or even incorrect column names. It seems pretty tough to fix that case.

Reply:
I will add comments for `csv()` in DataFrameReader and in readwrite.py as well in streaming.py